### ISSUE TITLE
Any good way to use categorical features for XGBoost models?

### ISSUE THREAD

**Issue by fernandoprocy** [o19s/elasticsearch-learning-to-rank#236]
Hi,

I'm currently working on a system that will require live resorting of query results obtained from ElasticSearch. I stumbled on this plugin and read the docs, and it seems like a good solution since for this first iteration we can settle with GBDT (XGBoost). However, some of the features that I use to train the model are categorical (think, for example, "movie genre"). Since it's a decision tree based model we have two main methods to deal with such type of category: by creating a map from an integer to the category or by using techniques like one-hot encoding (so the category variable will be encoded in N-column instead).

After reading the documentation I couldn't figure out if there is an easy way to handle these types of categories. Inside ElasticSearch the fields are stored as strings, so in order to use them for live resorting one would have to have at least a hashmap. Another solution would be to create a categorical field for every category, but this approach can quickly grow out of control, especially for features with reasonably high cardinality.

Is there any recommended way to deal with this issue? It appears to be a common issue since categorical variables are so common, but didn't see that in the docs or in here.

Thanks

**Comment by Onuchin-Artem** [comment]
you can use target encoding  to encode categorical data. I.e. instead of using movie genre itself as a feature use use click_rate of a genre. Then you can just index click_rate of a genre as elastic search field.

Reference: https://medium.com/datadriveninvestor/improve-your-classification-models-using-mean-target-encoding-a3d573df31e8

**Comment by fernandoprocy** [comment]
That's a good trick, thanks for sharing. I'm not a fan of using the label to derive features from. I think it helps only in the case where you have a test set that was generated from the same distribution of the train set. In a production scenario, I think the model may even perform worse due to it overfitting to the training data. Also, in the end, even that would require some sort of preprocessing and indexing.

IMO, the best way to handle these high cardinality categorical features would be to create an embedding for each category using the clickstream data (sessions) as the context. But unfortunately, this approach would require indexing one field per embedding dimension and that's one of the things I wanted to avoid.

I decided to remove the high cardinality features altogether from the first version, but if someone stumbles on a similar problem than mine, I hope it helps.
